Add timeout mechanism to Push function instead of retries

guoyuhong opened the pull request:
<!--
Thank you for your contribution!

Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->
Add timeout mechanism to Push function instead of retries. (As discussed with @eric-jj @raulchen @elibol )

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->
N/A

guoyuhong commented:
@elibol Please help to review this code change.

elibol commented on src/ray/object_manager/object_manager.h:
```diff
@@ -41,8 +41,11 @@ struct ObjectManagerConfig {
+  /// The time in milliseconds to wait until a Push request
+  /// fails due to unsatisfied local object. Special value:
+  /// Negative: waiting infinitely.
+  /// 0: failing directly.
```
Does this mean: If the object is not local, then fail immediately? If so, can you update to read: "Fail immediately if the object is not local."

robertnishihara replied:
"gives up" or "times out" would be a little clearer than "fails"

elibol commented on src/common/state/ray_config.h:
```diff
@@ -227,15 +227,18 @@ class RayConfig {
+  /// Timeout, in milliseconds, to wait until the Push request fails.
+  /// Special value:
+  /// Negative: waiting infinitely.
+  /// 0: failing directly.
```
Same comment applies here as in `object_manager.h`.

elibol commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+            [this, object_id, client_id](const boost::system::error_code &error) {
+              // Timer killing will receive the boost::asio::error::operation_aborted,
+              // we only handle the timeout event.
+              if (!error) {
```
Can you add a test that:
1. Forces `Push` to timeout by pushing an object that doesn't exist.
2. Invokes `Push` before an object is added locally to ensure the cancel mechanism functions correctly.

guoyuhong replied:
I will have a try.

elibol commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+    auto &clients = unfulfilled_push_tasks_[object_id];
+    if (clients.count(client_id) == 0) {
+      // If config_.push_timeout_ms < 0, we give an empty timer
+      // and the task will be kept infinitely.
```
It might make sense to add tests for these special configuration settings. Since the current tests do not support changing configuration, this is optional:
1. `config_.push_timeout_ms < 0`; A regular `Push` with this setting is sufficient.
2. `config_.push_timeout_ms == 0`; Do a regular `Push` and a `Push` with an object that doesn't exist.

If you don't have time to do this, I will eventually update the tests to support this and add this test at a later time.

guoyuhong replied:
I have finished adding tests to cover the code paths.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5658/
Test PASSed.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -90,6 +88,19 @@ void ObjectManager::NotifyDirectoryObjectAdd(const ObjectInfoT &object_info) {
+    for (auto &pair : iter->second) {
+      main_service_->post(
+          [this, object_id, pair]() { RAY_CHECK_OK(Push(object_id, pair.first)); });
+      if (pair.second != nullptr) {
```
When would `pair.second` equal `nullptr`?

EDIT: I see, this happens when the `timeout` equals `0`, right? In that case, I think we should just not add a timer.

guoyuhong replied:
This happens when `timeout < 0`, we need to keep this record in the map until the object notification calls it.

robertnishihara replied:
Oh I see, can you add a comment here explaining that?

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
-                                int retry) {
+void ObjectManager::HandlePushTaskTimeout(const ObjectID &object_id,
+                                          const ClientID &client_id) {
+  RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
```
`RAY_LOG(ERROR)` is too strong, this should probably be `RAY_LOG(DEBUG)` because this can happen for valid reasons during normal execution all of the time, right?

eric-jj replied:
I agree RAY_LOG(ERROR) is too strong, but I don't think it should be RAY_LOG(DEBUG) here. RAY_LOG(WARNING) may be a better one.

robertnishihara replied:
The issue is that `Push` is called in situations even when it isn't possible to perform the push, so this timeout will happen regularly during normal execution. This could result in the logs growing too verbose.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+      auto timer = std::shared_ptr<boost::asio::deadline_timer>();
+      if (config_.push_timeout_ms == 0) {
+        // The Push request fails directly when config_.push_timeout_ms == 0.
+        RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
```
`RAY_LOG(ERROR)` is too verbose for this, use `DEBUG`

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+              }
+            });
+      }
+      clients.emplace(client_id, timer);
```
When `config_.push_timeout_ms == 0`, there is no reason to add a null timer to `unfulfilled_push_tasks_`, right?

guoyuhong replied:
Yes, when `config_.push_timeout_ms == 0` we don't need to add this record. However,  when `config_.push_timeout_ms < 0`, we need to add this record, because -1 means waiting infinitely.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+  RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
+                 << " after waiting for " << config_.push_timeout_ms << " ms.";
+  auto iter = unfulfilled_push_tasks_.find(object_id);
+  if (iter != unfulfilled_push_tasks_.end()) {
```
Instead, can we do `RAY_CHECK(iter != unfulfilled_push_tasks_.end())`?

robertnishihara replied:
Both should happen on the `main_service_`, so it shouldn't be an issue (@elibol can you confirm?). If they could happen concurrently then we would be in trouble because they both modify `unfulfilled_push_tasks_`.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+                 << " after waiting for " << config_.push_timeout_ms << " ms.";
+  auto iter = unfulfilled_push_tasks_.find(object_id);
+  if (iter != unfulfilled_push_tasks_.end()) {
+    iter->second.erase(client_id);
```
We should do

```c++
num_erased = iter->second.erase(client_id);
RAY_CHECK(num_erased == 1);
```

guoyuhong replied:
The same as above comment. If that would happen, we shall not use `RAY_CHECK`.

robertnishihara replied:
Copying my reply from the previous comment 

Both should happen on the `main_service_`, so it shouldn't be an issue (@elibol can you confirm?). If they could happen concurrently then we would be in trouble because they both modify `unfulfilled_push_tasks_`.

elibol replied:
Yes, all invocations to `Push` execute on `main_service_`. We ensure this by running all asio sockets on `main_service_`. Only object transfers are executed on the send/receive thread pools.

guoyuhong commented:
@robertnishihara For the perf concern, this code change is for rare cases, in normal cases, it won't happen. Is that true? @elibol

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
+        // The Push request fails directly when config_.push_timeout_ms == 0.
+        RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
+                       << " due to direct failure setting. ";
+        return ray::Status::OK();
```
As a general style comment, I think it often improves clarity if functions have only a single `return` statement. I realize this function already has multiple return statements, but if we can reduce the number that would be great.

guoyuhong replied:
I can reduce one of my return statement by adding one if statement..

eric-jj commented on src/ray/object_manager/object_manager.h:
```diff
@@ -194,6 +196,13 @@ class ObjectManager : public ObjectManagerInterface {
+  /// The timer is for removing a push task due to unsatisfied local object.
+  std::unordered_map<
+      ObjectID,
+      std::unordered_map<ClientID, std::shared_ptr<boost::asio::deadline_timer>>>
```
can we use unique_ptr to replace shared_ptr here?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5686/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5705/
Test PASSed.

elibol commented:
@guoyuhong yes. I think @robertnishihara is thinking ahead to when we use `Push` for all objects, regardless of whether they're local or not during task forwarding (in `node_manager.cc`). We'll have to make sure this implementation does not cause performance issues when we make that change.

guoyuhong commented:
Hi @robertnishihara @elibol , I have added some tests in object_manager_test.cc. I added a new object_manager in that test using different configurations and added more Pull & Push calls in order to cover all the added code paths. I also added a Push call after waiting 2 seconds, so the object_manager_test consumes more time. Here is the 2-round  tests using this code change vs. old version. 
Round 1: 
Object Manger Stress Test: 2381 ms -> 2388 ms.
Object Manger Test: 2044 ms -> 5094 ms.
Round 2: 
Object Manger Stress Test: 2379 ms -> 2384 ms.
Object Manger Test: 2039 ms -> 5087 ms.
Is this acceptable?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5715/
Test FAILed.

robertnishihara commented:
@guoyuhong yes, increasing the duration of these tests by a little bit is completely fine.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5749/
Test PASSed.

guoyuhong commented:
Hi @robertnishihara @elibol , if there is no further concern, please help to sign off this pull request. Thank you!

robertnishihara commented:
@guoyuhong could you add the `RAY_CHECK` checks that I requested before merging this?

I'm ok with merging it with `RAY_LOG(WARNING)`. However, I strongly suspect that this will be too verbose and will need to be reverted to `RAY_LOG(DEBUG)`.

guoyuhong commented:
@elibol You mean use different derived classes for different scenarios and configurations instead of starting multiple servers, right?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5770/
Test PASSed.

guoyuhong commented:
Thanks @robertnishihara and @elibol for all your comments. I have removed the third server in the test and add `RAY_CHECK` as required.

robertnishihara commented:
Thanks!
